### PR TITLE
Implement verify signature syscall and cleanup

### PR DESCRIPTION
**Pull request by austinabell**
**Summary of changes**
Changes introduced in this pull request:
- Implements verify sig syscall #301 
  - Removes duplicate error when syscalls are called in actors
    - (I don't actually think the out of gas exit code can be hit within the actors and maybe that is intentional)
- Crypto crate cleanup
  - A lot was setup to try to match spec 1:1 and old implementation when possible, so now more closely matches impls and cleaner interface in rust
  - Signature verification now also returns result instead of just a boolean, for better error reporting

**Reference issue to close (if applicable)**

Closes 
#301

**Other information and links**

### CODE REVIEW

File: blockchain/chain_sync/src/sync.rs [hunk 1]
```diff
@@ -574,14 +573,12 @@ where
...
         if header.signature().is_none() {
-            return Err(Error::Validation("Signature is nil in header"));
+            return Err(Error::Validation("Signature is nil in header".to_owned()));
         }
 
         let base_tipset = self.load_fts(&TipSetKeys::new(header.parents().cids.clone()))?;
-        println!("base tipset {:?}", base_tipset);
         let parent_tipset = base_tipset.tipset()?;
 
-        println!("BEFORE TIME STAMP VALIDATIONS");
```
Comment (dutterbutter) [inline blockchain/chain_sync/src/sync.rs:584]: hehehe

File: crypto/src/signature.rs [hunk 2]
```diff
@@ -102,40 +102,53 @@ impl Signature {
...
+
+        // BLS verify hash against key
+        if verify(&sig, &[hashed], &[pk]) {
+            Ok(())
+        } else {
+            Err("bls signature verification failed".to_owned())
+        }
+    }
+
+    /// Returns true if a secp256k1 signature is valid
```
Comment (dutterbutter) [inline crypto/src/signature.rs:135]: Misleading comment as it does not return `bool`?

File: blockchain/blocks/src/header.rs [hunk 3]
```diff
@@ -268,11 +268,11 @@ impl BlockHeader {
         let signature = self
             .signature()
             .as_ref()
-            .ok_or(Error::InvalidSignature("Signature is nil in header"))?;
+            .ok_or_else(|| Error::InvalidSignature("Signature is nil in header".to_owned()))?;
```
Comment (ec2) [inline blockchain/blocks/src/header.rs:271]: There's nothing more beautiful in the world than a big fat method chain

Comment (austinabell) [reply blockchain/blocks/src/header.rs:271]: ????????

**Pull request merged**